File: 0144. Preorder Traversal/solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# Recursive Solution
class Solution(object):
    def preorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        if root is None: return []

        left = self.preorderTraversal(root.left)
        right = self.preorderTraversal(root.right)

        return [root.val] + left + right

File: 0144. Preorder Traversal/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

// Recursive Solution
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        if (!root) return vector<int> {};

        auto rootTraverse = vector<int> {root->val};
        auto leftTraverse = preorderTraversal(root->left);
        auto rightTraverse = preorderTraversal(root->right);

        rootTraverse.insert(rootTraverse.end(), leftTraverse.begin(), leftTraverse.end());
        rootTraverse.insert(rootTraverse.end(), rightTraverse.begin(), rightTraverse.end());

        return rootTraverse;
    }
};

// Iterative
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        if (!root) return vector<int> {};

        std::stack<TreeNode*> Stack;
        Stack.push(root);
        vector<int> res;

        while (!Stack.empty()){
            auto cur = Stack.top();
            res.push_back(cur->val);
            Stack.pop();

            if (cur->right) Stack.push(cur->right); 
            if (cur->left) Stack.push(cur->left);
        }

        return res;
    }
};